Use installed keg formula files when referencing installed formulae/dependencies

Rylan12 opened the pull request:
As promised in slack, this PR includes the changes necessary to finally `brew install` and `brew uninstall` using only the internal API.

To test, run:

```console
$ rm -rf "$(brew --cache)/api"

$ export HOMEBREW_USE_INTERNAL_API=1

$ brew install abcde
✔︎ JSON API formula.arm64_sequoia.jws.json
✔︎ JSON API cask.arm64_sequoia.jws.json
✔︎ JSON API abcde.json
==> Fetching downloads for: abcde
✔︎ Bottle Manifest abcde (2.9.3_1)
✔︎ Bottle Manifest cdrtools (3.02a09)
✔︎ Bottle cdrtools (3.02a09)
✔︎ Bottle Manifest eye-d3 (0.9.8)
✔︎ Bottle eye-d3 (0.9.8)
✔︎ Bottle Manifest glyr (1.0.10_3)
✔︎ Bottle glyr (1.0.10_3)
✔︎ Bottle Manifest libdiscid (0.6.5)
✔︎ Bottle libdiscid (0.6.5)
✔︎ Bottle Manifest mkcue (1)
✔︎ Bottle mkcue (1)
✔︎ Bottle Manifest berkeley-db@5 (5.3.28_1)
✔︎ Bottle Manifest gdbm (1.26)
✔︎ Bottle gdbm (1.26)
✔︎ Bottle Manifest perl (5.40.2)
✔︎ Bottle Manifest libao (1.2.2)
✔︎ Bottle libao (1.2.2)
✔︎ Bottle Manifest vorbis-tools (1.4.3)
✔︎ Bottle vorbis-tools (1.4.3)
✔︎ Bottle abcde (2.9.3_1)
✔︎ Bottle perl (5.40.2)
✔︎ Bottle berkeley-db@5 (5.3.28_1)
==> Installing dependencies for abcde: cdrtools, eye-d3, glyr, libdiscid, mkcue, berkeley-db@5, gdbm, perl, libao and vorbis-tools
==> Installing abcde dependency: cdrtools
==> Pouring cdrtools--3.02a09.arm64_sequoia.bottle.tar.gz
🍺  /opt/homebrew/Cellar/cdrtools/3.02a09: 210 files, 6.3MB
==> Installing abcde dependency: eye-d3
==> Pouring eye-d3--0.9.8.all.bottle.tar.gz
🍺  /opt/homebrew/Cellar/eye-d3/0.9.8: 151 files, 1014KB
==> Installing abcde dependency: glyr
==> Pouring glyr--1.0.10_3.arm64_sequoia.bottle.tar.gz
🍺  /opt/homebrew/Cellar/glyr/1.0.10_3: 18 files, 358.7KB
==> Installing abcde dependency: libdiscid
==> Pouring libdiscid--0.6.5.arm64_sequoia.bottle.tar.gz
🍺  /opt/homebrew/Cellar/libdiscid/0.6.5: 12 files, 160.0KB
==> Installing abcde dependency: mkcue
==> Pouring mkcue--1.arm64_sequoia.bottle.2.tar.gz
🍺  /opt/homebrew/Cellar/mkcue/1: 7 files, 83.7KB
==> Installing abcde dependency: berkeley-db@5
==> Pouring berkeley-db@5--5.3.28_1.arm64_sequoia.bottle.tar.gz
🍺  /opt/homebrew/Cellar/berkeley-db@5/5.3.28_1: 5,272 files, 86.4MB
==> Installing abcde dependency: gdbm
==> Pouring gdbm--1.26.arm64_sequoia.bottle.tar.gz
🍺  /opt/homebrew/Cellar/gdbm/1.26: 25 files, 1MB
==> Installing abcde dependency: perl
==> Pouring perl--5.40.2.arm64_sequoia.bottle.tar.gz
🍺  /opt/homebrew/Cellar/perl/5.40.2: 2,785 files, 69.8MB
==> Installing abcde dependency: libao
==> Pouring libao--1.2.2.arm64_sequoia.bottle.3.tar.gz
🍺  /opt/homebrew/Cellar/libao/1.2.2: 56 files, 332.5KB
==> Installing abcde dependency: vorbis-tools
==> Pouring vorbis-tools--1.4.3.arm64_sequoia.bottle.tar.gz
🍺  /opt/homebrew/Cellar/vorbis-tools/1.4.3: 20 files, 665.2KB
==> Installing abcde
==> Pouring abcde--2.9.3_1.arm64_sequoia.bottle.tar.gz
🍺  /opt/homebrew/Cellar/abcde/2.9.3_1: 295 files, 3.2MB
==> Running `brew cleanup abcde`...
==> No outdated dependents to upgrade!

$ brew uninstall abcde
==> Downloading https://formulae.brew.sh/api/cask.jws.json
Uninstalling /opt/homebrew/Cellar/abcde/2.9.3_1... (295 files, 3.2MB)
==> Autoremoving 10 unneeded formulae:
berkeley-db@5
cdrtools
eye-d3
gdbm
glyr
libao
libdiscid
mkcue
perl
vorbis-tools
Uninstalling /opt/homebrew/Cellar/glyr/1.0.10_3... (18 files, 358.7KB)
Uninstalling /opt/homebrew/Cellar/vorbis-tools/1.4.3... (20 files, 665.2KB)
Uninstalling /opt/homebrew/Cellar/mkcue/1... (7 files, 83.7KB)
Uninstalling /opt/homebrew/Cellar/perl/5.40.2... (2,785 files, 69.8MB)
Uninstalling /opt/homebrew/Cellar/libdiscid/0.6.5... (12 files, 160.0KB)
Uninstalling /opt/homebrew/Cellar/cdrtools/3.02a09... (210 files, 6.3MB)
Uninstalling /opt/homebrew/Cellar/eye-d3/0.9.8... (151 files, 1014KB)
Uninstalling /opt/homebrew/Cellar/libao/1.2.2... (56 files, 332.5KB)
Uninstalling /opt/homebrew/Cellar/gdbm/1.26... (25 files, 1MB)
Uninstalling /opt/homebrew/Cellar/berkeley-db@5/5.3.28_1... (5,272 files, 86.4MB)

$ tree "$(brew --cache)/api"
/Users/rylanpolster/Library/Caches/Homebrew/api
├── cask_names.txt
├── cask.jws.json
├── formula
│   ├── abcde.json
│   ├── autoconf.json
│   ├── automake.json
│   ├── berkeley-db@5.json
│   ├── bison.json
│   ├── ca-certificates.json
│   ├── cdrtools.json
│   ├── cmake.json
│   ├── dvdrtools.json
│   ├── eye-d3.json
│   ├── flac.json
│   ├── gdbm.json
│   ├── gettext.json
│   ├── gh.json
│   ├── glib.json
│   ├── glyr.json
│   ├── lame.json
│   ├── libao.json
│   ├── libdiscid.json
│   ├── libogg.json
│   ├── libtool.json
│   ├── libunistring.json
│   ├── libvorbis.json
│   ├── m4.json
│   ├── meson.json
│   ├── mkcue.json
│   ├── mpdecimal.json
│   ├── ninja.json
│   ├── openssl@3.json
│   ├── pcre2.json
│   ├── perl.json
│   ├── pkgconf.json
│   ├── python-setuptools.json
│   ├── python@3.12.json
│   ├── python@3.13.json
│   ├── readline.json
│   ├── smake.json
│   ├── sqlite.json
│   ├── texinfo.json
│   ├── vorbis-tools.json
│   └── xz.json
├── formula_aliases.txt
├── formula_names.txt
└── internal
    ├── cask.arm64_sequoia.jws.json
    └── formula.arm64_sequoia.jws.json

3 directories, 47 files
```

Importantly, the full `formula.jws.json` file was not downloaded during this process.

MikeMcQuaid commented on Library/Homebrew/cleanup.rb:
```diff
@@ -718,11 +718,14 @@ def self.autoremove(dry_run: false)
       Formula.clear_cache unless dry_run
 
-      formulae = Formula.installed
+      formulae = Formula.installed(prefer_stub: true)
```
Anything dealing with installed formulae should just be able to use the formula from the Cellar rather than a stub?

MikeMcQuaid commented on Library/Homebrew/formulary.rb:
```diff
@@ -1067,7 +1067,12 @@ def klass(flags:, ignore_errors:)
     sig { overridable.params(flags: T::Array[String]).void }
     def load_from_api(flags:)
-      json_formula = Homebrew::API::Formula.all_formulae[name]
+      json_formula = if Homebrew::EnvConfig.use_internal_api?
+        Homebrew::API::Formula.formula_json(name)
```
This definitely makes sense.

MikeMcQuaid commented on Library/Homebrew/installed_dependents.rb:
```diff
@@ -26,7 +26,7 @@ def find_some_installed_dependents(kegs, casks: [])
       # First, attempt to resolve the keg to a formula
       # to get up-to-date name and tap information.
-      f = keg.to_formula
+      f = keg.to_formula(prefer_stub: true)
```
Same deal with kegs: they should be able to use the formula file they have internally.

MikeMcQuaid reviewed (commented):
Thanks! A few comments. I think stubs should only be needed during the `brew install`/`brew upgrade`/maybe `brew reinstall` paths as most anything else will be either
1. triggering an installation
2. can use a formula from the Cellar

Rylan12 commented:
Okay, I've finally been able to rework this.

I've added `Formulary::from_installed` that loads a formula directly from the file in `/opt/homebrew/opt/NAME/.brew/NAME.rb`. I'm not thrilled with the naming of the method, so please suggest improvements.

This method differs from `Formulary::from_keg`, because the new method loads the formula file that was present at install time. So, any updates that have been made since will not be present in the `Formula` returned by `from_installed`.

To prevent potential issues, I've added new methods like `Formula#installed_formula_runtime_dependencies` that return these types of formulae, whereas `Formula#formula_runtime_dependencies` will still try to return the newest `Formula` available, regardless of what's installed.

One known issue is that running `brew install foo` will download JSON files for all of `foo`'s dependencies, even those that are already installed and up to date. Eventually, this will be improved using the minimal API, but I've left that off for now.

Also, the only place that `prefer_stub:` is used right now is in `named_args.rb`, which can probably be cleaned up in a follow-up PR.

@MikeMcQuaid is this approach more what you were picturing?

MikeMcQuaid reviewed (commented):
Thanks, yeh, this approach makes much more sense to me!

The only thing I'm not sure I understand is how/why the behaviour changes here compared to e.g. `from_rack` or `from_keg`?

Should e.g. `from_keg` perhaps be adjusted to make the changes you're looking for here?

MikeMcQuaid commented:
Thanks @Rylan12! Will look in more detail tomorrow and happy to help with test failures if you get blocked.



> Now, the difference between the `Formulary` methods is summarized as follows;

Sounds good.



> * `resolve`, `from_rack`, and `from_keg` attempt to load the installed version of a formula, and fall back to `factory`

`resolve` is the only one that maybe could/should default to Not From The Keg, if that's a behaviour change.

> Future work: improving `brew outdated` and `brew upgrade` with the internal API (it just made sense to push that to a second PR

Makes sense, thanks again!

Rylan12 commented:
> `resolve` is the only one that maybe could/should default to Not From The Keg, if that's a behaviour change.

@MikeMcQuaid `resolve` calls `from_rack` (and therefore `from_keg`), so it's a consequence of modifying the `from_keg` behavior. We can definitely avoid that, but it might mean it's better to leave `from_keg` as-is and add a new method like what I had originally called `from_installed`.

Are there specific cases that you're worried about with the modified `resolve`?

MikeMcQuaid reviewed (approved):
Looks good, thanks @Rylan12!

Rylan12 commented:
The `installed_dependents` checks are still failing, so I'm looking into it. Probably won't happen until tomorrow, though

Rylan12 commented:
Okay, as per usual, the fix ended up being incredibly simple, yet only apparent after days of troubleshooting 🤦

Anyway, I think this is ready now! I want to wait until just after the next release to merge in case there are issues
